lib/updatehub/updatehub.c

```diff
@@ -0,0 +1,920 @@
+		}
+
+		if (strcmp(sha256_image_dowloaded, _update_info.sha256sum_image) !=
+		    0) {
+			LOG_ERR("SHA256SUM of image are not the same");
+			_ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
+			goto cleanup;
+		}
+	}
+
+	_ctx.code_status = UPDATEHUB_OK;
+
+cleanup:
+	k_free(data);
+}
+
+static enum updatehub_response install_update(void)
+{
+	if (boot_erase_img_bank(DT_FLASH_AREA_IMAGE_1_ID) != 0) {
+		LOG_ERR("Failed to init flash and erase second slot");
+		_ctx.code_status = UPDATEHUB_FLASH_INIT_ERROR;
+		goto error;
+	}
+
+	if (tc_sha256_init(&_ctx.sha256sum) < 1) {
+		LOG_ERR("Could not start sha256sum");
+		_ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
+		goto error;
+	}
+
+	if (!start_coap_client()) {
+		_ctx.code_status = UPDATEHUB_NETWORKING_ERROR;
+		goto error;
+	}
+
+	if (coap_block_transfer_init(&_ctx.block, COAP_BLOCK_1024,
+				     _update_info.image_size) < 0) {
+		LOG_ERR("Unable init block transfer");
+		_ctx.code_status = UPDATEHUB_NETWORKING_ERROR;
+		goto cleanup;
+	}
+
+	flash_img_init(&_ctx.flash_ctx);
+
+	_ctx.downloaded_size = 0;
+	int verification_download = 0;
+	int attempts_download = 0;
+
+	while (_ctx.downloaded_size !=
+	       _ctx.block.total_size) {
+		verification_download = _ctx.downloaded_size;
+
+		if (send_request(COAP_TYPE_CON, COAP_METHOD_GET,
+				 UPDATEHUB_DOWNLOAD) < 0) {
+			_ctx.code_status = UPDATEHUB_NETWORKING_ERROR;
+			goto cleanup;
+		}
+
+		install_update_cb();
+
+		if (_ctx.code_status != UPDATEHUB_OK) {
+			goto cleanup;
+		}
+
+		if (verification_download == _ctx.downloaded_size) {
+			if (attempts_download == COAP_MAX_RETRY) {
+				LOG_ERR("Could not get the packet");
+				_ctx.code_status =
+					UPDATEHUB_DOWNLOAD_ERROR;
+				goto cleanup;
+			}
+			attempts_download++;
+		}
+	}
+
+cleanup:
+	cleanup_conection();
+
+error:
+	memset(&_update_info, 0, sizeof(_update_info));
+	_ctx.downloaded_size = 0;
+
+	return _ctx.code_status;
+}
+
+static int report(enum updatehub_state state)
+{
+	char *device_id = k_malloc(DEVICE_ID_SIZE);
+	char *firmware_version = k_malloc(BOOT_IMG_VER_STRLEN_MAX);
+	int ret = -1;
+
+	if (!device_identity_get(device_id)) {
+		goto error;
+	}
+
+	if (!firmware_version_get(firmware_version)) {
+		goto error;
+	}
+
+	struct report report;
+	const char *exec = state_name(state);
+
+	memset(&report, 0, sizeof(report));
+	report.product_uid = CONFIG_UPDATEHUB_PRODUCT_UID;
+	report.device_identity.id = device_id;
+	report.version = firmware_version;
+	report.hardware = CONFIG_BOARD;
+	report.status = exec;
+	report.package_uid = _update_info.package_uid;
+
+	switch (_ctx.code_status) {
+	case UPDATEHUB_INSTALL_ERROR:
+		report.previous_state =
+			state_name(UPDATEHUB_STATE_INSTALLING);
+		break;
+	case UPDATEHUB_DOWNLOAD_ERROR:
+		report.previous_state =
+			state_name(UPDATEHUB_STATE_DOWNLOADING);
+		break;
+	case UPDATEHUB_FLASH_INIT_ERROR:
+		state_name(UPDATEHUB_FLASH_INIT_ERROR);
+		break;
+	default:
+		report.previous_state = "";
+	}
```

Comment:
Does not make difference here but better to put break; after all the cases. There are some misra checks that will complain about these.